src/f0/app_solar/source/solar.c

```diff
@@ -38,83 +64,176 @@ static const DACConfig dac1cfg = {
 static INA226Driver ina226dev;
 
 /**
- * @brief control DAC output in millivolts.
- *
+ * @brief control DAC output in microvolts.
+ * @param[in] dacp      DAC driver pointer.
+ * @param[in] chan      DAC channel.
+ * @param[in] uv        output volts in uv (microVolts).
  */
-void dacPutMillivolts(DACDriver *dacp, dacchannel_t chan, uint32_t mv) {
+ 
+void dacPutMillivolts(DACDriver *dacp, dacchannel_t chan, uint32_t uv) {
     /* Per section 14.5.3 of the STM32F0x1 ref manual,
      * Vout(mV) = VDDA(mV) * (reg_val / 4096)
      * so, reg_val = (Vout * 4096) / VDDA
      */
-    dacsample_t val = (mv << 12) / DAC_VDDA_MV;
+    dacsample_t val = ((uv/100) << 12) / (DAC_VDDA_UV/100);
     dacPutChannelX(dacp, chan, val);
 }
 
+
 /**
  * @brief Calculates Iadj_v required for a particular max current.
- *
- * @return Iadj_v in millivolts.
- */
+ * @param[in] i_out      current in uA.
+ * @return    Iadj_v     DAC output in microvolts.
+ */    
 uint32_t calc_iadj(uint32_t i_out)
 {
-    return ((5052000 - i_out * RSENSE) / 3200);
+  /* Per page 7 of LT1618 datasheet
+   * V_Isense = (1.263 - 0.8* v_iadj)/25
+   * => 25 * i_out * Rsense  = 1.263 - 0.8 * V_iadj
+   * Multiplying 4000 on both sides
+   * 100000 * i_out * Rsense = 5052 - 3200 * V_iadj
+   * Now i_out is in uA, Rsense is in mOhm and V_iadj is in uV
+   * => 100000 * i_out/1000000 * Rsense/1000 = 5052 - 3200 * V_iadj/1000000
+   * => 100 * i_out * Rsense = 5052000000 - 3200 * V_iadj
+   * => i_out * Rsense = 50520000 - 32 * V_iadj
+   * => V_adj = (5052000 - i_out * Rsense * 100)/3200
+   */
+  return ((50520000 - i_out * RSENSE) / 32);
 }
 
-int32_t calc_mppt(uint16_t volt, int16_t curr, uint16_t pwr)
+/**
+ * @brief Maximum power point tracking algorithm for Solar cells
+ * @param[in] volt      Solar cell bus voltage in uV.
+ * @param[in] curr      Solar cell current in uA.
+ * @param[in] pwr       Power output from solar cells in microWatts.
+ * @return Maximum current to be drawn from solar cells in uA
+ * TODO: Find min/max limits of step size. Also check the units that work the best
+ *       Average a few samples to reduce noise
+ *       reduce max current limit immediatly if current is lower that the allowed current.(brownout condition)
+ *       increase max current limit immediatly if voltage is higher that expected.(shutdown condition)
+ *       
+ */
+int32_t calc_mppt(int32_t volt, int32_t curr, int32_t pwr)
 {
-    /* Amount to adjust iadj_v */
-    int32_t adjust;
-
     /* The values from the previous iteration of the loop */
-    static uint16_t prev_volt = 0;
-    static int16_t  prev_curr = 0;
-    static uint16_t prev_pwr = 0;
-    int16_t delta_v = volt - prev_volt;
-    int16_t delta_i = curr - prev_curr;
-    int16_t delta_p = pwr  - prev_pwr;
+    static int32_t prev_volt = 0;
+    static int32_t prev_curr = 0;
+    static int32_t prev_pwr = 0;
```

Comment:
Does not mention explicitly if this is signed or unsigned in datasheet but this is a result of multiplication of two signed registers. Seemingly implying this may be signed as well. And signed operation returns appropriate values so far.

Reply:
When it is a signed value, it designates the most significant bit as "SIGN" for the "Bit Name" part of the table (like shunt voltage). Typically power is expressed as a magnitude, so unsigned. In this case, it make sense that it would treat it as an unsigned value since it is multiplying two 15 bit (in magnitude anyway) values together to give the result. If it was signed it would be much more likely to overflow that register. I'm pretty sure it treats it as unsigned in this device.

Reply:
I am not very sure but leaning to agree towards your conclusion on power register. The equation 4 would result in -ve number but MSB in datasheet does not specify most significant bit as sign bit on page 15. None of my runs has returned a negative power although I have got a very small -ve current. So, I believe you are correct on this one.